haplotagphase command (previously extend command)

nkkarpov opened the pull request:
The command for extending information from haplotagged reads to the VCF file has two parameters:
```-g```, which controls the percentage of reads that must agree on the allele for phasing to be assigned.
```-c```, which sets the threshold for the length of homopolymers at which the assignment should be turned off.

mrvollger commented on whatshap/cli/extend.py:
```diff
@@ -0,0 +1,206 @@
+                pysam.AlignmentFile(
+                    alignment_file,
+                    reference_filename=reference if reference else None,
+                    require_index=True,
```
If this is kept, maybe add a command line arg threads and pass that on to the pysam threads argument to allow for increased decompression speed.

mrvollger commented:
@marcelm or @schrins is there a way to access the HP and PS tag using the whatshap read object? Would save an extra pass over the bam file.

marcelm reviewed (commented):
The HP and PS tags would need to be stored within the C++ `Read` class. There’s some precedence for this: The BX tag is also stored, so I think I would be ok to do the same for the HP and PS tags. It’s some extra memory and a little bit of extra parse time when constructing a Read object, but it is probably dwarfed by everything else that WhatsHap does. Search for `BX_tag` in the codebase to see how it is handled. If you’re not comfortable editing the C++ code, I can also help.

I will be unavailable for the rest of this week.

mrvollger commented:
That looks like a way I can help. If you'd be willing to incorporate it @nkkarpov I'll make another branch and try and add the code that exposes HP and PS in the read class?

nkkarpov commented on whatshap/cli/extend.py:
```diff
@@ -0,0 +1,206 @@
+                        if alignment.has_tag("PS") and alignment.has_tag("HP"):
+                            reads_to_ht[alignment.qname] = (
+                                int(alignment.get_tag("PS")) - 1,
+                                int(alignment.get_tag("HP")) - 1,
```
Yes, you are right. Probably, we also need to report a warning in that case.

nkkarpov commented:
I think it is ok if you can push changes to my branch directly. I've added you as a collaborator.

mrvollger commented:
@nkkarpov I added `has_{HP,PS}_tag` and `get_{HP,PS}_tag` functions for the Read class to your branch, and I think I did it all corectly though I am not experienced with Cython. At a minimum it passes the same number of tests as before I added the code on my machine.

I also synced the branch with main, since I initially did the work on my own fork which was up to date with main.

nkkarpov commented on whatshap/variants.py:
```diff
@@ -261,6 +261,16 @@ def _alignments_to_reads(self, alignments, variants, sample, reference):
+                hp = -1
+            
+            try:
+                ps = int(alignment.bam_alignment.get_tag("PS"))
```
It's a good question! I wrote this command assuming that I am working with the output of WhatsHap. However, I think making PS a string would be better, as it would not cause errors in parts of the code that don't rely on the assumption that this tag is an integer.

mrvollger replied:
I'd be okay with that as long as there isn't a standard we should follow. Maybe we can get some help. 

@tobiasmarschall do you know what type PS "should" be or who we might ask before we switch to string?

mrvollger replied:
Went through the hiphase code and Matt uses an int there.

tobiasmarschall replied:
The VCF specs say `PS` is an int (see Table 2 in https://samtools.github.io/hts-specs/VCFv4.3.pdf). So I'm in favor of ensuring it is an int (and giving an appropriate error message if it's not).

mrvollger replied:
Done! Thanks for pointing me to the vcf spec @tobiasmarschall. I was checking the sam tag spec...

nkkarpov commented:
The pull request at https://github.com/whatshap/whatshap/pull/517 includes code that merges supplementary alignments with primary alignments. Depending on which one is merged first, this code should be modified in one of two pull requests.

nkkarpov commented:
@mrvollger I think the current version should give reasonable assignments of PS and HP to VCF entries, at least on the data I tested. I'm going to test more, but if you have some data on where you were going to test, I would be happy to hear feedback.

marcelm commented:
Can you rebase on top of main? Let me know if you need help with the test failures. Otherwise, I’ll review as soon as the tests pass.

marcelm commented on whatshap/cli/extend.py:
```diff
@@ -0,0 +1,238 @@
+                if max_length >= cut_poly:
+                    continue
+        super_reads[0].append(Variant(pos, allele=al1, quality=score1))
+        super_reads[1].append(Variant(pos, allele=al1 ^ 1, quality=score1))
```
XORing with 1 is the same as NOT. Is the intention to map 1 to 0 and 0 to 1? Then please use `1 - al1`. Or `int(not al1)` maybe.

marcelm reviewed (commented):
Here’s a partial review.

I’m not so fond of the name `extend`, it doesn’t feel like there is any "extending" going on. `extend` sounds more as if this were something like imputation. This subcommand takes existing phasing information and applies it to a VCF. Isn’t this the opposite direction of what `haplotag` does? Maybe the name should also be accordingly.

Brainstorming: `apply`, `applyphase`, `maptag`, `mapphase`, `haplomap`, `haplotagphase`.

tobiasmarschall commented:
Out of @marcelm's suggestions, I'd vote for `haplotagphase`.

nkkarpov commented:
I also like `haplotagphase`. Let me provide some insight into how we utilize this command. At present, it integrates into our workflow as follows:

```bash
whatshap phase --only-snvs --reference reference.fasta -o phased-snv.vcf.gz unphased.vcf.gz alignments.bam
tabix -p vcf phased-snv.vcf.gz
whatshap haplotag --reference reference.fasta -o haplotagged.bam phased-snv.vcf.gz alignments.bam
samtools index haplotagged.bam
whatshap extend --reference reference.fasta -o phased.vcf.gz phased-snv.vcf.gz haplotagged.bam
```

In essence, this process allows us to extend phasing from phased single nucleotide variants (SNVs) to all types of variants. The current approach is slightly unconvinient, as it necessitates passing large BAM files through the pipeline multiple times. Moving forward, I want to enable haplotyping to be performed on-the-fly within this command, thus making `haplotagphase` more indicative of its function.

nkkarpov commented:
Hi @marcelm, I addressed all the comments and renamed the command to ```haplotagphase```. Additionally, I have added some words into ```CHANGES.rst``` and ```guide.rst```.

marcelm commented:
Hi, I was away at a work-related meeting. I should have time for a review next week. Feel free to remind me if I haven’t gotten back to you by Wednesday!

marcelm commented:
> In essence, this process allows us to extend phasing from phased single nucleotide variants (SNVs) to all types of variants. The current approach is slightly unconvinient, as it necessitates passing large BAM files through the pipeline multiple times. Moving forward, I want to enable haplotyping to be performed on-the-fly within this command, thus making `haplotagphase` more indicative of its function.

I’m not entirely on board. I see this new subcommand as an inverse of `haplotag`. While `haplotag` takes existing phasing information from a VCF and "transfers" it to the reads in a BAM file, this new command would take phasing information from a BAM and transfer that to the variants in a VCF.

What you’re describing is a bit different because the input would be a partially phased VCF, and I don’t think merging `haplotagphase` as it is now and then making it work for the above scenario is the right approach.

IIUC, your desired outcome is that 1) `phase` does not use indels for making phasing decisions but 2) indels *are* phased in the output (using the surrounding variants). However, I’d say this functionality should be part of `phase`.

I believe I remember that you have already tried modifying `phase`, so I’m not quite sure what I’m misunderstanding.

mrvollger commented:
Not contributing much here but I wanted say that my use case for this is the simpler case where I have reads phased using another method and I want to phase a vcf. So I think the command is still of a lot of value as is even if it doesn't solve this indel problem.

marcelm commented on whatshap/cli/haplotagphase.py:
```diff
@@ -0,0 +1,351 @@
+    length.
+
+    Args:
+        - only_indels (bool): If True, only consider indels for inclusion in the consensus. SNVs are ignored.
```
There should be no leading dashes, see [example](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html#overview).

marcelm commented on doc/guide.rst:
```diff
@@ -1186,3 +1187,27 @@ It can be used as follows:
+Given a haplotagged BAM file, a file with variants, and a reference, this command sequence outputs a phased VCF file::
+
+
+    tabix -p vcf input.vcf.gz
```
The `-p vcf` isn’t necessary (it seems to be autodetected).

marcelm reviewed (commented):
Thanks once again, and here’s another round of review. Most are minor issues, except that I would like to see the test case extended a little bit to verify that the output actually contains phased variants, and I think there were two minor logic issues.

I didn’t have any more time to work through `consensus()`, but in case I don’t find the time and nothing else majors comes up, I’m fine with merging the PR as soon as my comments from this round are addressed.

mrvollger commented:
FYI @nkkarpov, I ran this on some of my own data, and the "IGV" test looks good to me so far.

nkkarpov commented on whatshap/cli/haplotagphase.py:
```diff
@@ -0,0 +1,351 @@
+        if ht < 0 or ps < 0:
+            continue
+        if ht > 1:
+            number_of_skipped += 1
```
yes, you are right. I fixed it.

nkkarpov commented on tests/test_run_haplotagphase.py:
```diff
@@ -0,0 +1,45 @@
+        alignment_file="tests/data/pacbio/haplotagged.bam",
+        reference="tests/data/pacbio/reference.fasta",
+        output="/dev/null",
+    )
```
I checked that I phased most of the variants. This occurs because, for some variants, we observe a tie that we do not wish to break with the default parameters. For example, if we set the threshold for votes to 0, we assign only 1 variant vs 3 in `tests/data/pacbio/phased.vcf`.

nkkarpov commented:
I have a feeling that the logic around the phase is a little bit complicated, but yeah, it make sense to make this postprocessing in phase command (with some additional flag). I need to think about it more.

I addressed all comments, and currently, only tests for MacOS fail (which is also true for main branch).

marcelm commented:
This looks fine to me now. Any further thoughts or is this ready to be merged?

(We need to ignore the macOS test failure for now.)

mrvollger commented:
Nothing from me!

nkkarpov commented:
Nothing from my side.

marcelm commented:
Awesome, let’s do this, and thanks for the contribution!

mrvollger commented:
Awesome @marcelm! Do you have an estimate of when the next release might be?

marcelm commented:
I’ve released 2.3 now.

tobiasmarschall commented:
Thanks you all! :fire:
